tests/kohana/LogTest.php

```diff
@@ -61,51 +61,360 @@ public function test_attach_attaches_log_writer_and_returns_this()
+	{
+		// initialize
+		$logger = new Log;
+		$writer = new Kohana_LogTest_Log_Writer_Memory();
 		$logger->attach($writer);
 
-		$this->assertSame($logger, $logger->detach($writer));
+		// test logging with specialized method
+		$logger->$method($message); /* Record line number for fuzzy test */ $expected_line = __LINE__;
+		$logger->flush();
+		$expected = $writer->logs[0];
 
-		$this->assertAttributeSame(array(), '_writers', $logger);
+		// reset writer's logs
+		$writer->logs = array();
+
+		// test logging with the generic log method
+		foreach ($levels as $level) {
+			$method = 'log';
+			$logger->$method($level, $message); /* Record line number */ $actual_line = __LINE__;
+		}
+		$logger->flush();
+
+		// assertions
+		$line_delta = $actual_line - $expected_line;
+		foreach ($writer->logs as $actual) {
+			$this->assertLogMessageEquals($expected, $actual, $line_delta, 1);
+		}
 	}
 
+	/**
+	 * Tests that Log casts message to string
+	 *
+	 * @test
+	 */
+	public function test_log_casts_message_to_string()
+	{
+		// initialize
+		$logger = new Log;
+		$writer = new Kohana_LogTest_Log_Writer_Memory();
+		$logger->attach($writer);
+
+		// test logging with an object as message
+		$object = new Kohana_LogTest_ToString();
+		$logger->log(\Psr\Log\LogLevel::INFO, $object);
+		$logger->flush();
+		$actual = $writer->logs[0];
 
+		// assertions
+		$this->assertInternalType('string', $object->message);
+		$this->assertInternalType('string', $actual['body']);
+		$this->assertSame($object->message, $actual['body']);
+	}
+
+	/**
+	 * Provider for test_immediate_flushing
+	 */
+	public function provider_immediate_flushing()
+	{
+		return [
+			[
+				TRUE, // flush immediately
+				1, // $expected_number_of_logs
+			],
+			[
+				FALSE, // do NOT flush immediately
+				0, // $expected_number_of_logs
+			]
+		];
+	}
+
+	/**
+	 * Tests that Log flushes immediately when immediate flushing is ON, otherwise
+	 * it delays flushing after call to flush()
+	 *
+	 * @test
+	 * @dataProvider provider_immediate_flushing
+	 */
+	public function test_immediate_flushing($flush_immediately, $expected_number_of_logs)
+	{
+		// initialize
+		$logger = new Log;
+		$writer = new Kohana_LogTest_Log_Writer_Memory();
+		$logger->attach($writer);
+
+		// Turn on immediate flushing
+		$logger->set_immediate_flush($flush_immediately);
+
+		// test logging with an object as message
+		$logger->log(\Psr\Log\LogLevel::DEBUG, 'dummy message');
+
+		// We will NOT call $logger->flush() here...
+
+		$actual_number_of_logs = count($writer->logs);
+
+		// assertions
+		$this->assertSame($flush_immediately, $logger->get_immediate_flush());
+		$this->assertSame($expected_number_of_logs, $actual_number_of_logs);
+
+		// force flush and assert anew
+		$logger->flush();
+		$expected_number_of_logs = 1; // expect logs after forced flush
+		$actual_number_of_logs = count($writer->logs); // recalculate actual
+		$this->assertSame($expected_number_of_logs, $actual_number_of_logs);
```

Comment:
`$this->assertSame(1, count($writer->logs), 'Expect logs written after forced flush');` would be more readable in code and produce a more obvious failure exception message if it fails IMO.